src/equations/equations.jl

```diff
@@ -124,6 +125,47 @@ end
   return flux
 end
 
+# Dirichlet-type boundary condition for use with UnstructuredQuadMesh
+#  Note: For unstructured we lose the concept of a "direction"
+@inline function (boundary_condition::BoundaryConditionDirichlet)(u_inner, orientation, x, t,
+                                                                  surface_flux_function, equation)
+  # get the external value of the solution
+  u_boundary = boundary_condition.boundary_value_function(x, t, equation)
+
+  # Calculate boundary flux
+  flux = surface_flux_function(u_inner, u_boundary, orientation, equation)
+
+  return flux
+end
+
+"""
+    BoundaryConditionWall(boundary_wall_function)
+
+Create a generic wall type boundary condition that uses the function `boundary_wall_function`
+to specify the external solution values based upon the solution internal values.
+
+# Example
+```julia
+julia> BoundaryConditionWall(free_slip_wall_state)
+```
+
+!!! warning "Experimental code"
+    This boundary condition can change any time and is currently only implemented for the
+    CompressibleEulerEquations2D
+"""
+struct BoundaryConditionWall{B}
+  boundary_wall_function::B
+end
+
+@inline function (boundary_condition::BoundaryConditionWall)(u_inner, orientation, x, t,
+                                                             surface_flux_function, equations)
+  u_boundary = boundary_condition.boundary_wall_function(u_inner, orientation, equations)
+
+  flux = surface_flux_function(u_inner, u_boundary, orientation, equations)
+
+  return flux
+end
```

Comment:
```suggestion
@inline function (boundary_condition::BoundaryConditionWall)(u_inner, normal::AbstractVector, x, t,
                                                             surface_flux_function, equations)
  u_boundary = boundary_condition.boundary_wall_function(u_inner, normal, equations)

  flux = surface_flux_function(u_inner, u_boundary, normal, equations)

  return flux
end
```
This is basically the same implementation as the `BoundaryConditionDirichlet`, isn't it? If so, I am against repeating the same code again.

Reply:
Basically, except it gets the value of `u_boundary` from a different external state routine. I suppose we can have `BoundaryConditionDirichlet` and `BoundaryConditionWall` both contain a routine called `boundary_value_function` and then it should be obvious to the user what kind of boundary function this is based on the boundary condition type name

Reply:
What I meant is: `BoundaryConditionWall(some_function)` behaves exactly as `BoundaryConditionDirichlet(some_function)` whenever `some_function` is the same, isn't it?

Reply:
I see, thanks for the clarification! It would indeed be good to describe the expected signature of the functions in the docstrings. Then, feel free to resolve this thread.